chain/shard/tx_queue.go

```diff
@@ -0,0 +1,93 @@
+package shard
+
+import (
+	"github.com/argus-labs/world-engine/chain/x/shard/types"
+	"github.com/zyedidia/generic/queue"
+	"sync"
+)
+
+// TxQueue acts as a transaction queue. Transactions come in to the TxQueue with at tick.
+type TxQueue struct {
+	lock       sync.Mutex
+	ntx        NamespacedTxs
+	outbox     []*types.SubmitCardinalTxRequest
+	moduleAddr string
+}
+
+type txQueue struct {
+	// tickQueue are the ticks waiting to be submitted to the blockchain
+	tickQueue *queue.Queue[uint64]
+	// txs are the transaction requests, indexed by tick.
+	txs map[uint64]*types.SubmitCardinalTxRequest
+}
+
+// NamespacedTxs maps namespaces to a transaction queue.
+type NamespacedTxs map[string]*txQueue
+
+func (tc *TxQueue) TxsForNamespaceInTick(ns string, tick uint64) *types.SubmitCardinalTxRequest {
+	return tc.ntx[ns].txs[tick]
+}
+
+// AddTx first checks if there are already transactions stored for the tick in the request.
+// If there are, we simply append this request to txs.
+// If there aren't yet, we append the tick number to tickQueue, then append to the txs map.
+func (tc *TxQueue) AddTx(namespace string, tick, txID uint64, payload []byte) {
+	tc.lock.Lock()
+	defer tc.lock.Unlock()
+	// handle the case where this is a brand-new namespace
+	if tc.ntx[namespace] == nil {
+		tc.ntx[namespace] = &txQueue{
+			tickQueue: queue.New[uint64](),
+			txs:       make(map[uint64]*types.SubmitCardinalTxRequest),
+		}
+		tc.ntx[namespace].tickQueue.Enqueue(tick)
+		tc.ntx[namespace].txs[tick] = &types.SubmitCardinalTxRequest{
+			Sender:    tc.moduleAddr,
+			Namespace: namespace,
+			Tick:      tick,
+			Txs:       &types.Transactions{Txs: []*types.Transaction{{TxId: txID, SignedPayload: payload}}},
+		}
+		return
```

Comment:
I think this function should be reorganized a bit. Remove this return, and limit this if branch to only set up the namespace. Use the tickQueue.Empty() branch and nil checks below to actually add the transactions. This way, the logic for adding stuff is only implemented in one place.

Reply:
updated, lmk how it looks now

Reply:
👍